### ISSUE TITLE
Issues with Python 3.6

### ISSUE THREAD

**Issue by m3drano** [alecthomas/injector#59]
I just updated from Python 3.5.2 to 3.6.0 and I'm having some issues with injector. It seems that when using injector with the new annotation syntax, Python 3.6 introduces some changes that make @injector.inject incompatible. See an example stack trace here:

```
  File "${HOME}/${PROJECT}/${SCRATCH}/execroot/simulation/bazel-out/local-py3-fastbuild/bin/main.runfiles/__main__/main.py", line 15, in main
    runner()
  File "${HOME}/${PROJECT}/simulation/simulation.py", line 140, in runner
    simulator = custom_injector.get(Simulation)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 707, in get
    result = scope_instance.get(key, binding.provider).get(self)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 142, in get
    return injector.create_object(self._cls)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 751, in create_object
    (), additional_kwargs, e, self._stack,)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 71, in reraise
    raise exception.with_traceback(tb)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 744, in create_object
    init(instance, **additional_kwargs)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 1079, in inject
    kwargs=kwargs
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 843, in call_with_injection
    owner_key=self_.__class__ if self_ is not None else callable.__module__,
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 58, in wrapper
    return function(*args, **kwargs)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 886, in args_to_inject
    instance = self.get(key.interface)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 707, in get
    result = scope_instance.get(key, binding.provider).get(self)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 58, in wrapper
    return function(*args, **kwargs)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 546, in get
    provider = InstanceProvider(provider.get(self.injector))
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 142, in get
    return injector.create_object(self._cls)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 751, in create_object
    (), additional_kwargs, e, self._stack,)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 71, in reraise
    raise exception.with_traceback(tb)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 744, in create_object
    init(instance, **additional_kwargs)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 1079, in inject
    kwargs=kwargs
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 843, in call_with_injection
    owner_key=self_.__class__ if self_ is not None else callable.__module__,
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 58, in wrapper
    return function(*args, **kwargs)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 886, in args_to_inject
    instance = self.get(key.interface)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 707, in get
    result = scope_instance.get(key, binding.provider).get(self)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 58, in wrapper
    return function(*args, **kwargs)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 546, in get
    provider = InstanceProvider(provider.get(self.injector))
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 142, in get
    return injector.create_object(self._cls)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 751, in create_object
    (), additional_kwargs, e, self._stack,)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 71, in reraise
    raise exception.with_traceback(tb)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 744, in create_object
    init(instance, **additional_kwargs)
  File "${HOME}/${PROJECT}/simulation/activity_distribution.py", line 343, in __init__
    do_merge=False, trace_file='trace_file')
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 1079, in inject
    kwargs=kwargs
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 843, in call_with_injection
    owner_key=self_.__class__ if self_ is not None else callable.__module__,
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 58, in wrapper
    return function(*args, **kwargs)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 886, in args_to_inject
    instance = self.get(key.interface)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 707, in get
    result = scope_instance.get(key, binding.provider).get(self)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 142, in get
    return injector.create_object(self._cls)
  File "${HOME}/${VIRTUALENV}/lib/python3.6/site-packages/injector.py", line 728, in create_object
    instance = cls.__new__(cls)
  File "/usr/local/Cellar/python3/3.6.0/Frameworks/Python.framework/Versions/3.6/lib/python3.6/typing.py", line 1126, in __new__
    return _generic_new(cls.__next_in_mro__, cls, *args, **kwds)
  File "/usr/local/Cellar/python3/3.6.0/Frameworks/Python.framework/Versions/3.6/lib/python3.6/typing.py", line 1095, in _generic_new
    obj.__init__(*args, **kwds)
TypeError: __init__() missing 2 required positional arguments: 'injector' and 'target'
```

The Python binary comes from Homebrew and it is running on macOS 10.12.2, although that should not be a problem. The exact same code runs smoothly on Python 3.5.2. I'm trying to debug further.

**Comment by jstasiak** [comment]
This is caused by a typing change[1] that introduces a behaviour unexpected by Injector (a Generic subclass' ``__new__`` automatically calls its ``__init__`` now), I'll see if it can be changed easily, alternatively I'll report it as a potential bug.

[1] https://github.com/python/typing/commit/50d2b4d377043a5b6c268b532c13e2e2bfe61352

**Comment by jstasiak** [comment]
This was caused by some internal typing changes between typing 3.5.2.2 and 3.5.3; the relevant issues are resolved in https://github.com/alecthomas/injector/commit/0793123bbc5e7d2649171e778d1f70b6e7384a76 and will be released shortly.

**Comment by jstasiak** [comment]
This just got released in 0.12.0, sorry about the delay.
